### DISCUSSION TITLE
Raspberry Pi 4 Compute Module support

### DISCUSSION THREAD

**Discussion by ryao** [NVIDIA/open-gpu-kernel-modules#239] [Ideas]
https://www.jeffgeerling.com/blog/2022/external-graphics-cards-work-on-raspberry-pi

There are issues with the BCM2711's PCIe root complex that make it difficult to get graphics cards working on the Raspberry Pi.

It would be extremely cool if some workarounds could be added to this driver to support the Raspberry 4 Compute Module.

I do not expect this to be a priority of any kind. I am putting the idea out there for those interested in nerd credit. :)

**Comment by stalkerg** [comment]
@ryao honestly, I believe will be much easier to wait for Raspberry Pi 5 with fixed PCIe behavior, currently, it's just broken in some aspects. Broadcom should take attention here.

**Comment by ApplePie420** [comment]
It would be neat to try on other ARM boards with PCIe capabilities, such as ROCKpro64 (RK3399 SoC), that *may* have the PCIe configured properly. It kinda sucks that RPi's almost always had their stuff somehow limited/bodged.

**Comment by Qinka** [comment]
Will the Raspberry Pi 4 be any better with the UEFI firmware?

**Comment by stalkerg** [comment]
@Qinka no, the main problem is that Raspberry Pi 4 not fully complain to PCIe standard by design.

**Comment by arpieb** [comment]
So.... It's 2025, and the Pi 5 is out with PCIe support.  Just tried to hook up the following bits:

- Raspberry Pi 5 8GB
- NVIDIA Titan Xp GPU
- GeeekPi P02 PCIe Slot for Raspberry Pi 5
- [ATX Power Management Hat for Pi 5](https://www.tindie.com/products/tomtibbetts/atx-power-management-hat-for-pi-5/)
- ATX 500W PSU

The goal is to put the GPU online so that I could host something like Ollama or some other straight-to-GPU solutions that just need a lightweight interface and network connectivity on the cheap, which the Pi 5 provides.  Inspiration was [this article](https://alican-kiraz1.medium.com/run-llm-on-pi5-connecting-an-nvidia-gpu-to-raspberry-pi-5-via-pcie-x4-a6d52c3efd2a), which nobody appears to be able to replicate.

I'm getting errors from the driver trying to initialize the card, and the [good folks on the NVIDIA forum](https://forums.developer.nvidia.com/t/installing-titan-xp-on-raspberry-pi-5-via-pcie-board-getting-rminitadapter-failed/324106) said it appears to be because the proprietary drivers don't support the Broadcom PCI bridge (vendor ID 0x14E4) based on this debug output:

```
Feb 17 22:40:56 pi5 kernel: [drm] [nvidia-drm] [GPU ID 0x00000100] Loading driver
Feb 17 22:40:56 pi5 kernel: [drm] Initialized nvidia-drm 0.0.0 20160202 for 0000:01:00.0 on minor 2
Feb 17 22:41:29 pi5 kernel: NVRM: Chipset not recognized (vendor ID 0x14e4, device ID 0x2712)
Feb 17 22:41:33 pi5 kernel: NVRM: GPU 0000:01:00.0: RmInitAdapter failed! (0x31:0xffff:2640)
Feb 17 22:41:33 pi5 kernel: NVRM: GPU 0000:01:00.0: rm_init_adapter failed, device minor number 0
```

I haven't tried the OSS drivers yet, but it sounds like as they stand they probably won't support it either out of the box.  How hard would it be to add?

Thanks!
